### PR TITLE
feature/p256

### PR DESCRIPTION
**Pull request by decanus**

### CODE REVIEW

File: contracts/algorithms/EllipticCurveAlgorithm.sol [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    * @param data The signed data to verify.
+    * @param signature The signature to verify.
+    * @return True iff the signature is valid.
+    */
+    function verify(bytes calldata key, bytes calldata data, bytes calldata signature) external view returns (bool) {
+        return curve.validateSignature(sha256(data), parseSignature(signature), parseKey(key));
+    }
+
+    function parseSignature(bytes memory data) internal view returns (uint256[2] memory) {
+        return [data.readUint256(0), data.readUint256(32)];
```
Comment (Arachnid) [inline contracts/algorithms/EllipticCurveAlgorithm.sol]: We already have `readBytes32`; could you use that instead, and cast the result?

File: package.json [hunk 2]
```diff
@@ -7,6 +7,7 @@
     "@ensdomains/buffer": "^0.0.6",
     "@ensdomains/solsha1": "^0.0.2",
     "dnsprovejs": "0.0.9",
+    "elliptic-solidity": "git+https://github.com/tdrerup/elliptic-curve-solidity.git",
```
Comment (Arachnid) [inline package.json]: Is it possible to pin this to a commit hash? I'd rather not have a dependency on solidity code that can be changed without notice.

**Comment by decanus** [comment]
**TODO:** Test with https://github.com/BANKEX/CurveArithmetics see gas difference.

File: contracts/algorithms/P256Algorithm.sol [hunk 3]
```diff
@@ -0,0 +1,35 @@
+pragma solidity ^0.5.0;
+
+import "./Algorithm.sol";
+import "../BytesUtils.sol";
+import "elliptic-solidity/contracts/curves/EllipticCurve.sol";
+
+contract P256Algorithm is Algorithm {
```
Comment (Arachnid) [inline contracts/algorithms/P256Algorithm.sol]: Sorry, actually this should be `SHA256P256Algorithm`.

File: test/data/algorithms.js [hunk 4]
```diff
@@ -56,5 +56,36 @@ module.exports = [
...
             "0x91108e1fabbb974406cbdaa90bd975b0b9dc25c38a14b27b1a18943a26eee2d798a79544f519dcae24a164dcfce66c2532034469c1582bf94fb4f89560fe1bc2"
         ]
-    ]
+    ],
+    [
+        "SHA256P256Algorithm",
+        // This test vector generated from the example in RFC5702 using the following Python script:
+        // import dns.rrset
+        //
+        // dnskey = dns.rrset.from_text("example.net.", 3600, "IN", "DNSKEY", "256 3 8 AwEAAcFcGsaxxdgiuuGmCkVImy4h99CqT7jwY3pexPGcnUFtR2Fh36BponcwtkZ4cAgtvd4Qs8PkxUdp6p/DlUmObdk=")
```
Comment (baloo) [inline test/data/algorithms.js]: This happens to be testing for rsa keys (the `8` in `256 3 8` is for rsa signature). You may want to steal examples from https://tools.ietf.org/html/rfc6605#section-6.1 for ecdsa p256 instead

Comment (baloo) [reply test/data/algorithms.js]: Here is script with samples from rfc6605:
``` python
from StringIO import StringIO
import dns.rrset

dnskey = dns.rrset.from_text("example.net.", 3600, "IN", "DNSKEY", "256 3 13 GojIhhXUN/u4v54ZQqGSnyhWJwaubCvTmeexv7bR6edb krSqQpF64cYbcB7wNcP+e+MAnLr+Wi9xMWyQLc8NAA==")

a = dns.rrset.from_text("www.example.net.", 3600, "IN", "A", "192.0.2.1")
buf = StringIO()
a.to_wire(buf)

rrsig = dns.rrset.from_text("www.example.net.", 3600, "IN", "RRSIG", "A 13 3 3600 20100909100439 20100812100439 55648 example.net.  qx6wLYqmh+l9oCKTN6qIc+bw6ya+KJ8oMz0YP107epXA yGmt+3SNruPFKG7tZoLBLlUzGGus7ZwmwWep666VCw==")
signature = rrsig[0].signature
rrsig[0].signature = ''
signdata = rrsig[0].to_digestable() + buf.getvalue()

print ('0x' + dnskey[0].to_digestable().encode('hex'), '0x' + signdata.encode('hex'), signature.encode('hex'))
```

Comment (Arachnid) [reply test/data/algorithms.js]: Good catch, thank you!

Comment (decanus) [reply test/data/algorithms.js]: yeah the comment was never updated, I was lazy

File: test/data/algorithms.js [hunk 5]
```diff
@@ -56,5 +56,37 @@ module.exports = [
...
+        // signature = rrsig[0].signature
+        // rrsig[0].signature = ''
+        // signdata = rrsig[0].to_digestable() + buf.getvalue()
+        //
+        // print ('0x' + dnskey[0].to_digestable().encode('hex'), '0x' + signdata.encode('hex'), signature.encode('hex'))
+        [
+            // example.net. 3600 IN DNSKEY 257 3 13 (
+            //         GojIhhXUN/u4v54ZQqGSnyhWJwaubCvTmeexv7bR6edb
+            //         krSqQpF64cYbcB7wNcP+e+MAnLr+Wi9xMWyQLc8NAA== )
+            "0x0101030d1a88c88615d437fbb8bf9e1942a1929f28562706ae6c2bd399e7b1bfb6d1e9e75b92b4aa42917ae1c61b701ef035c3fe7be3009cbafe5a2f71316c902dcf0d00",
```
Comment (baloo) [inline test/data/algorithms.js]: Is this supposed to be the wire format of the dns packet? It hasn't changed along with the new dnskey afaict.

Comment (decanus) [reply test/data/algorithms.js]: @baloo These are correct values, only the comment was incorrect.

Comment (baloo) [reply test/data/algorithms.js]: alright!

**Pull request merged**